[CALCITE-5982] Allow overloading the created enumerable in CalciteMetaImpl when calling getTables() or getColumns()

olivrlee opened the pull request.

wnob commented on core/src/main/java/org/apache/calcite/jdbc/CalciteMetaImpl.java:
```diff
@@ -263,23 +288,30 @@ private static ImmutableMap.Builder<DatabaseProperty, Object> addProperty(
+   * */
+  private String[] getColumnNames(Class<?> clazz) {
+    try {
+      Method m = clazz.getMethod("getColumnNames");
```
No need for dynamic method lookup with the type changes.

olivrlee replied:
I tried this modification: `getColumnNames(Class<? extends CalciteMetaTable> clazz){...}`  however it that `clazz` object still doesn't have direct access to the interface function getColumnNames()


Also, if this change worked, I would need 2 functions for this, one for `extends CalciteMetaTable` and one for `extends MetaColumn`
Thoughts?

olivrlee commented on core/src/main/java/org/apache/calcite/schema/Table.java:
```diff
@@ -82,4 +85,20 @@ public interface Table {
+  /**
+   * Overload to provide additional column metadata if available.
+   */
+  default Map<String, Map<String, Object>> getColumnMetadata() {
```
Yeah I agree an untyped map is not great. 

Would you say this modeling is still required within Calcite, if I go ahead with the factory interface?
Calcite in this case would not need to be aware of metadata at all. The Calcite factory implementation would just create `CalciteMetaTable`s

When a user of Calcite implements the factory interface,  the implementation would then be responsible for how the metadata is accessed, enumerated, typed, etc. and maybe that modeling should sit there.

barrkel replied:
I think, if a factory interface removes the need for the loosely typed map, then don't add the map at all.

olivrlee replied:
Yeah I like the factory approach, thanks for the review and suggestion.

I pushed a commit with the rough idea but there are still some tests to update/ and I need to fix the plugin retrieval, so it's not ready for re-review yet.

olivrlee commented:
@barrkel  Thanks for the factory suggestion, I have cleaned up this PR to use factories and updated the tests. Please take a look when you have time!

tanclary commented on core/src/main/java/org/apache/calcite/config/CalciteConnectionConfig.java:
```diff
@@ -110,4 +110,16 @@ public interface CalciteConnectionConfig extends ConnectionConfig {
   boolean lenientOperatorLookup();
```
nit: remove "/" from PR title?

tanclary commented on core/src/main/java/org/apache/calcite/jdbc/CalciteMetaColumnFactoryImpl.java:
```diff
@@ -0,0 +1,98 @@
+      String isNullable,
+      String isAutoincrement,
+      String isGeneratedcolumn) {
+    return new MetaColumn(
```
Not sure if outdated but looks like there are some constructor errors ?

olivrlee replied:
There's something I need to update, seems like Avatica main took out the last 2 parameters in the constructor. I haven't rebased my Avatica snapshot off of main, I'll do that

tanclary commented on core/src/main/java/org/apache/calcite/jdbc/CalciteMetaImpl.java:
```diff
@@ -90,13 +90,34 @@
         .setTransactionIsolation(Connection.TRANSACTION_NONE);
     this.connProps.setDirty(false);
+
+    if (metaTableFactory != null) {
```
Anytime I see this kind of `if` block I wish Java had the elvis operator

olivrlee replied:
Indeed

julianhyde replied:
Au contraire. The solution is to squash out the nulls.

The elvis operator makes it too easy to build horrible chains of optionals.

tanclary commented on core/src/main/java/org/apache/calcite/jdbc/CalciteMetaImpl.java:
```diff
@@ -263,23 +289,16 @@ private static ImmutableMap.Builder<DatabaseProperty, Object> addProperty(
+        metaTableFactory.getColumnNames());
   }
 
+
```
nit: remove blank line

tanclary commented on core/src/main/java/org/apache/calcite/jdbc/CalciteMetaImpl.java:
```diff
@@ -378,6 +375,7 @@ Enumerable<MetaTable> tables(String catalog) {
             tables(schema, Functions.<String>truePredicate1()));
   }
 
+
```
same as above

tanclary commented on core/src/test/java/org/apache/calcite/test/JdbcFrontJdbcBackTest.java:
```diff
@@ -73,6 +79,72 @@ class JdbcFrontJdbcBackTest {
+                connection.getMetaData().getTables(
+                    null, null, null, null);
+            assertTrue(rset.next());
+            assertEquals(rset.getMetaData().getColumnCount(), 11);
```
Clarification: where do these expected results come from ?

tanclary reviewed (commented):
I have limited knowledge on this but I think it looks great : )

tanclary commented:
I think some of the code smells on this might be helpful suggestions. Like marking some of the variables as final or protected. Of course maybe they're irrelevant.

julianhyde commented on core/src/main/java/org/apache/calcite/jdbc/CalciteMetaColumnFactory.java:
```diff
@@ -0,0 +1,48 @@
+      String isNullable);
+
+  /** Returns the list of expected column names. */
+  String[] getColumnNames();
```
Let's make this `List<String>`. Then you can return an `ImmutableList` and don't have to worry about making a defensive copy.

I'd give it a default implementation that returns the columns in the JDBC spec, and add the javadoc 'The default implementation returns the columns in the JDBC specification.'

julianhyde commented on core/src/main/java/org/apache/calcite/jdbc/CalciteMetaColumnFactory.java:
```diff
@@ -0,0 +1,48 @@
+  String[] getColumnNames();
+
+  /** Returns the class of MetaColumn that is created. */
+  Class<?> getMetaColumnClass();
```
'Returns the type of object created. Must be a sub-class of `MetaColumn`.'

Does it work if you change the return type from `Class<?>` to `Class<? extends MetaColumn>`?

julianhyde commented on core/src/main/java/org/apache/calcite/jdbc/CalciteMetaImpl.java:
```diff
@@ -90,13 +90,40 @@
+    this(connection, new CalciteMetaTableFactoryImpl(), new CalciteMetaColumnFactoryImpl());
+  }
+
+  public CalciteMetaImpl(CalciteConnectionImpl connection,
```
Add javadoc.

I would remove the `@Nullable`. Caller can easily provide factories.

olivrlee replied:
It's the Java 11 CI that is giving me issues when I remove the @nullable annotation. 

https://github.com/apache/calcite/actions/runs/6255263301/job/16984285491?pr=3421

julianhyde replied:
I think CI is finding cases where you might accidentally supply null (e.g. when `Driver` calls `config.metaColumnFactory`). I'll fix.

julianhyde commented on core/src/main/java/org/apache/calcite/jdbc/CalciteMetaImpl.java:
```diff
@@ -178,11 +205,16 @@ private <E> MetaResultSet createResultSet(Enumerable<E> enumerable,
+        } catch (NoSuchFieldException e2) {
+          throw new RuntimeException(e2);
+        }
       }
```
I am surprised that `getDeclaredField` would ever succeed if `getField` has failed.

olivrlee replied:
I was testing on my subclasses and `getField()` returns the 10 fields of the default `MetaTable`  class attributes, and `getDeclaredField()` returns the additional properties on my subclass.

julianhyde commented on core/src/main/java/org/apache/calcite/jdbc/CalciteMetaColumnFactory.java:
```diff
@@ -0,0 +1,48 @@
+      String typeName,
+      Integer columnSize,
+      @Nullable Integer decimalDigits,
+      Integer numPrecRadix,
```
If `numPrecRadix` nullable? If not, make it `int`. If so, add `@Nullable`.

julianhyde commented on core/src/main/java/org/apache/calcite/jdbc/CalciteMetaTableFactory.java:
```diff
@@ -0,0 +1,32 @@
+  MetaTable newMetaTable(Table table, String tableCat, String tableSchem, String tableName);
+
+  /** Returns the list of expected column names. */
+  String[] getColumnNames();
```
Switch from `String[]` to `List<String>` so that you don't need defensive copy.

julianhyde commented on core/src/test/java/org/apache/calcite/test/JdbcFrontJdbcBackTest.java:
```diff
@@ -73,6 +79,75 @@ class JdbcFrontJdbcBackTest {
+  /**
+   * Sample subclass used in {@link JdbcFrontJdbcBackTest#testTablesExtraColumn()}.
+   */
+  public static class MetaExtraTable extends CalciteMetaTable {
```
Does this class need to be public? Does its field need to be public?

If it can be private, that's good. If it needs to be public (because, say, reflection) then you should say that it needs to be public in the javadoc.

julianhyde commented on core/src/test/java/org/apache/calcite/test/JdbcFrontJdbcBackTest.java:
```diff
@@ -73,6 +79,75 @@ class JdbcFrontJdbcBackTest {
+
+    @Override public String[] getColumnNames() {
+      // 11 columns total.
+      return new String[] {
```
Would it be an error if the columns were not prefixed by the list of standard columns? (e.g. adding an extra column in the middle, or removing a column, or re-ordering columns)

Figure out what the necessary constraints are, and add a test that checks that those constraints are being enforced.

julianhyde reviewed (commented):
Nice work. The interface looks very clean.

My suggested changes are largely cosmetic, and I hope they're easy to apply.

olivrlee commented on core/src/main/java/org/apache/calcite/jdbc/CalciteMetaColumnFactory.java:
```diff
@@ -0,0 +1,48 @@
+/** Factory for creating MetaColumns for getColumns(). */
+public interface CalciteMetaColumnFactory {
+  /** Instantiates a MetaColumn. */
+  MetaColumn newMetaColumn(
```
oh sorry I missed this last week somehow. Let me update right now
